### PR TITLE
[kafka/internal, kafkaexporter, kafkareceiver] Add SASL mechanism "AWS_MSK_IAM_OAUTHBEARER" to kafkaexporter

### PR DESCRIPTION
**Pull request by donald-cheung**
**Description:** 
This PR added the SASL mechanism "AWS_MSK_IAM_OAUTHBEARER" to kafkaexporter and kafkareceiver. This mechanism use the AWS MSK IAM SASL Signer for Go https://github.com/aws/aws-msk-iam-sasl-signer-go. This mechanism is added because the "AWS_MSK_IAM" is not working in our cluster and also in this [issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19747). We added an new mechanism instead of replace the existing one because we want to keep the backward compatibility just in case someone is using "AWS_MSK_IAM".

**Link to tracking Issue:** 
[19747](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19747)

**Testing:** 
We built the images and tested the SASL mechanism in our team.
We added related unit tests.

**Documentation:** 
We updated the kafkaexporter and kafakreciever README on the SASL mechanism.

### CODE REVIEW

**Comment by mx-psi** [comment]
@donald-cheung Thanks for your contribution! We need to you to sign the CLA in order to be able to accept your contribution

**Comment by donald-cheung** [comment]
@mx-psi 
Thanks for the reminder. I am asking my company to sign the CCLA so it will take some time.

**Comment by mattiasnensen** [comment]
Hey, this would be awesome to have. What's going on?

**Comment by donald-cheung** [comment]
Sorry, we are in the process of getting approval from the upper organization. This will take some time. Thank you.

**Comment by donald-cheung** [comment]
We will have a meeting and update the progress next week. 
BTW, how can I reopen this PR? There is no button in the bottom. Thank you.

**Comment by mx-psi** [comment]
@donald-cheung Let me know when you have approval from your org and want to reopen the PR and I can do it for you :)

**Comment by donald-cheung** [comment]
@mx-psi Thanks. The meeting is rescheduled to next week so I will update you later.

**Comment by canhnt** [comment]
Hi @donald-cheung, could you please keep posted on the result? This MR is useful for us.

**Comment by donald-cheung** [comment]
Sorry for the late reply. The feedback from the upper organization is good. The upper organization will start their procedures to approve.

**Comment by jnavarro86** [comment]
@donald-cheung I just tested it and it is working for me, Thanks.

**Comment by donald-cheung** [comment]
@mx-psi @MovieStoreGuy
May you reopen the PR please? My organization will sign the CLA and I want to fix the merge conflict. Thank you.

**Comment by donald-cheung** [comment]
@mx-psi, @MovieStoreGuy, @dmitryax Thank you for waiting. The CLA is signed.

**Comment by donald-cheung** [comment]
@mx-psi, @MovieStoreGuy, @dmitryax. Is there update on reopening this PR? Thank you.

**Comment by donald-cheung** [comment]
`build-and-test / checks (pull_request)` is not passing. I tried running `make genotelcontribcol` but there is nothing to commit.
```
> make genotelcontribcol
/Users/donald.cheung/Desktop/test-open-telemetry/opentelemetry-collector-contrib/.tools/builder --skip-compilation --config cmd/otelcontribcol/builder-config.yaml --output-path cmd/otelcontribcol
Flag --output-path has been deprecated, use config distribution::output_path
2024-08-06T18:22:54.127+0900	INFO	internal/command.go:125	OpenTelemetry Collector Builder	{"version": "", "date": "unknown"}
2024-08-06T18:22:54.130+0900	INFO	internal/command.go:161	Using config file	{"path": "cmd/otelcontribcol/builder-config.yaml"}
2024-08-06T18:22:54.130+0900	INFO	builder/config.go:142	Using go	{"go-executable": "/opt/homebrew/bin/go"}
2024-08-06T18:22:54.139+0900	INFO	builder/main.go:100	Sources created	{"path": "cmd/otelcontribcol"}
2024-08-06T18:22:54.957+0900	INFO	builder/main.go:191	Getting go modules
2024-08-06T18:22:55.313+0900	INFO	builder/main.go:107	Generating source codes only, the distribution will not be compiled.
/Applications/Xcode.app/Contents/Developer/usr/bin/make --no-print-directory -C cmd/otelcontribcol fmt
Makefile:4: warning: overriding commands for target `lint'
../../Makefile.Common:199: warning: ignoring old commands for target `lint'
gofmt  -w -s ./
/Users/donald.cheung/Desktop/test-open-telemetry/opentelemetry-collector-contrib/.tools/goimports -w  -local github.com/open-telemetry/opentelemetry-collector-contrib ./
> git status
On branch kafka-exporter-aws-iam-oauth-bearer
Your branch is up to date with 'origin/kafka-exporter-aws-iam-oauth-bearer'.

nothing to commit, working tree clean
```

**Comment by mx-psi** [comment]
@donald-cheung This is because of the toolchain directive https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c25c058422e0874a1868285e9658218a1f4ddbf2/cmd/otelcontribcol/go.mod#L7 Your local go version does not match that of CI. That's fine but you need to revert the change on the `toolchain` directive to what it was before

**Comment by donald-cheung** [comment]
@mx-psi Thank you.

**Comment by donald-cheung** [comment]
@mx-psi, @dmitryax, @MovieStoreGuy, @pavolloffay, @jpkrohling. This PR is ready for review. Thank you very much.

**Comment by mx-psi** [comment]
@pavolloffay @MovieStoreGuy PTAL!

**Comment by donald-cheung** [comment]
@pavolloffay @MovieStoreGuy Please let me know if there is anything to follow up. Thank you.

File: .chloggen/kafka-exporter-aws-iam-oauth-bearer.yaml
Comment (pavolloffay) [inline .chloggen/kafka-exporter-aws-iam-oauth-bearer.yaml:7]: isn't receiver missing here?

Comment (pavolloffay) [reply .chloggen/kafka-exporter-aws-iam-oauth-bearer.yaml:7]: I guess we can remove the internal/kafka?

**Review by pavolloffay — COMMENTED** [review]
LGTM,

is there a way to add some functional test?

File: .chloggen/kafka-exporter-aws-iam-oauth-bearer.yaml
Comment (donald-cheung) [reply .chloggen/kafka-exporter-aws-iam-oauth-bearer.yaml]: Thanks. I updated the changelog.

**Comment by donald-cheung** [comment]
> LGTM,
> 
> is there a way to add some functional test?

No simple way, a AWS MSK is needed for testing.

**Comment by donald-cheung** [comment]
@MovieStoreGuy  PTAL and let me know if  there is anything to follow up. Thank you.

File: internal/kafka/authentication.go
Comment (MovieStoreGuy) [inline internal/kafka/authentication.go:58]: Please pass in the context to avoid using `context.TODO`

**Review by MovieStoreGuy — COMMENTED** [review]
Sorry, one minor nit and that is it from me.

**Comment by MovieStoreGuy** [comment]
> @MovieStoreGuy PTAL and let me know if there is anything to follow up. Thank you.

Sorry it has taken me a bit, there is only one outstanding nit from me @donald-cheung

File: internal/kafka/authentication.go
Comment (donald-cheung) [reply internal/kafka/authentication.go:60]: I have changed the code to pass the context from parent's function. PTAL. Thank you.

**Comment by amal-v** [comment]
Waiting for this PR to be merged! 🙇

**Comment by mx-psi** [comment]
@MovieStoreGuy Could we get this merged if ready?

**Comment by jacastello** [comment]
If I can add any more support to this PR, my team built and tested this locally against MSK, and it works as expected. Really looking forward to a mainline release!

**Comment by donald-cheung** [comment]
@MovieStoreGuy, @pavolloffay
Please let me know if there is anything leave before it can be merged. Thank you.

**Comment by jalola** [comment]
Our team is also looking for this feature, hope it can be merged soon 🙇

**Comment by donald-cheung** [comment]
@mx-psi can you run the workflows again? Thanks.

**Comment by donald-cheung** [comment]
Hello @mx-psi @pavolloffay @MovieStoreGuy, do you have any schedule to merge this PR? Thank you.

**Comment by amal-v** [comment]
Waiting so long for on this PR. Can we get some traction here ? @jpkrohling

**Comment by donald-cheung** [comment]
Hello @mx-psi @pavolloffay @MovieStoreGuy, can I get some updates on the schedule to merge this PR? Thank you.

**Comment by golem1987** [comment]
Waiting for merging, need this option

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by MovieStoreGuy** [comment]
I have a few environment questions for you @donald-cheung ,

How was the performance of using the new auth provider? Did it cause any additional resource usage on the kafka cluster?
When I originally looked at IAM auth, the team and I roughly saw a significant increase on load due to the new auth, I was wondering if your tests showed the same.

**Comment by donald-cheung** [comment]
> I have a few environment questions for you @donald-cheung ,
> 
> How was the performance of using the new auth provider? Did it cause any additional resource usage on the kafka cluster? When I originally looked at IAM auth, the team and I roughly saw a significant increase on load due to the new auth, I was wondering if your tests showed the same.

Hello, can you provide more details on your test? We are trying to see if the increase on load also happens on our side.

**Pull request merged**
